### DISCUSSION TITLE
Moving away from Laravel Mix

### DISCUSSION THREAD

**Discussion by jeffreyvr** [tailpress/tailpress#51] [Ideas]
In version 1.0.0 a `theme` subfolder was created, placing `header.php`, `footer.php`, `single.php` in that folder. To make that work, functions like `tailpress()->get_header()` ([example](https://github.com/jeffreyvr/tailpress/blob/master/theme/index.php#L1])) etc. were introduced.

This was all done in order get around the [infinite loop issue](https://github.com/tailwindlabs/tailwindcss/issues/4691) that was introduced with Tailwind 2.2.0.

In short, we could no longer watch for files using wild cards, like this:

```
purge: {
    content: [
        './*/*.php',
        './**/*.php',
        './resources/css/*.css',
        './resources/js/*.js',
        './safelist.txt'
    ],
},
```

But needed to do this instead (every other root file would have to be added):
```
purge: {
    content: [
        './template-parts/*.php',
        './resources/css/*.css',
        './resources/js/*.js',
        './comments.php',
        './header.php',
        './footer.php',
        './footer.php',
        './single.php',
        './index.php',
        './404.php',
        './safelist.txt'
    ],
},
```
We also need to restart the watch proces every time we added a new root file to the list. Not ideal.

The theme-subfolder fixes this, but after using this structure during the development of a theme, I notice I run into issues, particularly when working with plugins like WooCommerce. Some plugins call for the core `get_header` and `get_footer` function, and with this setup - this will not work.

In addition, it adds more code to the boilerplate. In hindsight, I think I shouldn't have added the subfolder at all. I'd like this setup to adhere to WordPress core functions and not go overboard on custom stuff. Anyone is free to add that of course, but the boilerplate should remain minimal in that regard.

Another way to solve the infinite loop issue, is to not use webpack / Laravel Mix. Tailwind comes with it's own CLI tool and we can use that instead.

However, we still need to use something to bundle our Javascript. Instead of using Tailwind CLI and Laravel Mix (for JS only), we can use a bundler like esbuild.

I often like to run a watch script that watched for changes on my CSS as well as JS. By using the `concurrently` package, we can start up both the Tailwind and esbuild watch scripts. See: https://github.com/jeffreyvr/tailpress/blob/tailwindcli/package.json#L14

Please take a look at the setup with Tailwind CLI and esbuild. I want to release this soon, but feedback is appreciated. 

https://github.com/jeffreyvr/tailpress/tree/tailwindcli

Side note: I still love Laravel Mix. It has certainly made dealing with webpack much easier and I will still be using it on other projects.

**Comment by merijnponzo** [comment]
Hi jeffrey,

I tried the tailwind cli but then, trying to get vue running etc etc.. laravel-mix is just easier
would this be an option? Or is this just like your subfolder setup :-) 
I never had any problems with this approach

With my previous webpack config i always seperate the wordpress files within a new folder

![Screenshot 2021-09-08 at 23 48 47](https://user-images.githubusercontent.com/23075300/132590309-9fb45653-f446-4746-b4d5-1ff2a7cfe389.png)

**Comment by jeffreyvr** [comment]
This looks like the subfolder setup. Problems arise when you start adding theme support for WooCommerce for example. Since you don't have an header.php and footer.php in the root - functions like `get_header` etc. won't work as expected. So you then have shop pages without a header and footer and need to additional modification. I can imagine other plugins that provide templates would cause issues as well. 

Did you run into these kinds of issues?

It would be great if we could somehow make Laravel Mix / Webpack to support PostCSS glob. Tried a couple of things but nothing did the trick.

Side note: the current implementation is based on the advice from the Tailwind docs:
https://tailwindcss.com/docs/just-in-time-mode#styles-rebuild-in-an-infinite-loop

**Comment by jeffreyvr** [comment]
Was thinking about this some more. The file that causes the infinite loop issue with Laravel Mix / webpack, is `mix-manifest.json`. 

We could disable the creation of that file and that stops the infinite loop (just did a quick test and it seems to work).

Previously we used `mix-manifest.json` for cache busting, we can use something else for that instead. Not sure if disabling the file has additional drawbacks.

**Comment by merijnponzo** [comment]
hmm when using content: ["./app/*.php"] i also end up in an infinite loop, I guess since my dist folder is copied / or fonts/ or mix-manifest.json etc the watch is triggered, i was expecting it would not since i'm only targeting at *.php

I ended up using, just like you mentioned in the beginning of this discussion
```
 purge: {
    content: [
      "./header.php",
      "./footer.php",
      "./views/*.php",
      "./views/*.twig",
      "./app/**/*.php",
      "./views/**/*.twig",
    ],
  },
  ```
Since i use timber/twig most of my files are in the views folder, so i don't expect to change the content array a lot


my laravel mix:
```
mix.js("resources/js/app.js", "dist/js");
mix.js("resources/js/app-vue.js", "dist/js").vue({ version: 3 });
mix.setPublicPath("./");
// tailwind setup
mix.postCss("resources/css/app.css", "dist/css", [
  require("postcss-import"),
  require("tailwindcss"),
  require("rfs"),
  require("postcss-nested"),
  require("postcss-preset-env")({ stage: 0 }),
]);


if (mix.inProduction()) {
  mix.version();
  mix.then(() => {
    const convertToFileHash = require("laravel-mix-make-file-hash");
    convertToFileHash({
      publicPath: "./",
      manifestFilePath: "mix-manifest.json",
      delOptions: { force: true },
    });
  });
} else {
  mix.browserSync({
    watch: true,
    reloadDelay: 100,
    proxy: { target: "eyecanseeart.local", ws: true },
    files: ["*.php", "resources/*/"],
  });
}
```

It's a work in progress i use
```
function load_manifest(){
	$css = get_template_directory_uri(). mix_any( '/dist/css/app.css', get_stylesheet_directory(), 'mix-manifest.json');
	wp_enqueue_style( 'tailwind', $css);
}

add_action( 'wp_enqueue_scripts', 'load_manifest' );
```
to pull in the hashed files from the distfolder copied with 
```
laravel-mix-make-file-hash
```
and creates hashed files like app.1ee00ca6d13eaf68eb82.css

mix_any comes from 
https://github.com/mindkomm/theme-lib-mix

I use your function @jeffreyvr/tailwindcss-tailpress to pull in the colors from theme.json

That said for tailwindcss-tailpress, i would split up the project with a starter theme, 
and clear documentation about you can use @jeffreyvr/tailwindcss-tailpres within your tailwind config to pull in the theme.json colors and typography

Thanks!

**Comment by jeffreyvr** [comment]
Thanks for your response @merijnponzo!

I think I have a [working example](https://github.com/jeffreyvr/tailpress/tree/mix) now without the infinite loop issue occurring.

Will have to test it some more. And, then still need to figure out if reverting back to Mix is the way forward or to stick with the current setup of esbuild / tailwind cli.
